src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,225 @@
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var eventArgsTypeSymbol = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemEventArgs);
+
+                context.RegisterOperationBlockStartAction(context => AnalyzeOperationBlockStart(context, eventArgsTypeSymbol));
+            });
+        }
+
+        private static void AnalyzeOperationBlockStart(OperationBlockStartAnalysisContext context, INamedTypeSymbol? eventArgsTypeSymbol)
+        {
+            if (context.OwningSymbol is not IMethodSymbol methodSymbol ||
+                methodSymbol.IsOverride ||
+                methodSymbol.Parameters.Length == 0 ||
+                methodSymbol.IsImplementationOfAnyInterfaceMember() ||
+                methodSymbol.HasEventHandlerSignature(eventArgsTypeSymbol) ||
+                !context.Options.MatchesConfiguredVisibility(Rule, methodSymbol, context.Compilation,
+                    defaultRequiredVisibility: SymbolVisibilityGroup.All) ||
+                context.Options.IsConfiguredToSkipAnalysis(Rule, methodSymbol,
+                    context.Compilation))
+            {
+                return;
+            }
+
+            var typesUsedPerParameter = PooledConcurrentDictionary<IParameterSymbol, PooledConcurrentSet<ITypeSymbol>>.GetInstance(SymbolEqualityComparer.Default);
+            foreach (var param in methodSymbol.Parameters)
+            {
+                typesUsedPerParameter.AddOrUpdate(param, PooledConcurrentSet<ITypeSymbol>.GetInstance(SymbolEqualityComparer.Default), (param, value) => value);
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                var paramReference = (IParameterReferenceOperation)context.Operation;
+
+                switch (paramReference.Parent.Kind)
+                {
+                    case OperationKind.Invocation:
+                        AddDeclaringTypeSymbol(((IInvocationOperation)paramReference.Parent).TargetMethod);
+                        break;
+
+                    case OperationKind.PropertyReference:
+                    case OperationKind.EventReference:
+                    case OperationKind.FieldReference:
+                    case OperationKind.MethodReference:
+                        AddDeclaringTypeSymbol(((IMemberReferenceOperation)paramReference.Parent).Member);
+                        break;
+
+                    case OperationKind.Argument:
+                        AddType(paramReference.Type);
+                        break;
+
+                    case OperationKind.Conversion:
+                        AddType(paramReference.Parent.Type);
+                        break;
+
+                    default:
```

Comment:
How about the following?
```cs
void M(Derived d)
{
   Derived d2 = d;
   d2.SomeDerivedTypeMethod();
}
```

We probably need to identify cases where the parameter _escapes_ our tracking logic and we have to bail out on it.

Reply:
I am adding a test for that but I believe we should know about this change of reference (maybe not).

Reply:
I confirm that we do not report on this case.

Reply:
How about following:

```csharp
void M(Derived d1, Derived d2)
{
   d1.SomeBaseTypeMethod();
   d2.SomeDerivedTypeMethod();
}
```

Do we suggest changing the parameter type to Base for d1?

Reply:
> I confirm that we do not report on this case.

Hmm I guess that is because there are no other parameter references. How about the following slightly modified example:

```csharp
void M(Derived d)
{
   d.SomeBaseTypeMethod();

   Derived d2 = d;
   d2.SomeDerivedTypeMethod();
}
```

Reply:
We also do not report here.